mudata/_core/mudata.py

```diff
@@ -404,137 +425,229 @@ def _update_attr(self, attr: str, axis: int, join_common: bool = False):
+                data_mod,
             )
-            data_mod = data_mod.join(data_common, how="left", sort=False)
+
+            # Make simple attrmaps with anndata positions
+            # TODO: any edgecases?
+            mdict = dict()
+            incr = 0
+            for m in self.mod.keys():
+                mdict[m] = np.zeros(data_mod.shape[0])
+                mod_size = getattr(self.mod[m], attr).shape[0]
+                mdict[m][incr : (incr + mod_size)] = np.arange(mod_size) + 1
+                incr += mod_size
+
+        #
+        # General case: with duplicates and/or intersections
+        #
         else:
-            dfs = [
-                _make_index_unique(
-                    getattr(a, attr)
-                    .assign(**{rowcol: np.arange(getattr(a, attr).shape[0])})
-                    .add_prefix(m + ":")
+            if join_common:
+                # If all modalities have a column with the same name, it is not global
+                columns_common = reduce(
+                    np.intersect1d, [getattr(self.mod[mod], attr).columns for mod in self.mod]
+                )
+                data_global = data_global.loc[
+                    :, [c not in columns_common for c in data_global.columns]
+                ]
+                dfs = [
+                    _make_index_unique(
+                        getattr(a, attr)
+                        .drop(columns_common, axis=1)
+                        .assign(**{rowcol: np.arange(getattr(a, attr).shape[0])})
+                        .add_prefix(m + ":")
+                    )
+                    for m, a in self.mod.items()
+                ]
+
+                # Here, attr_names are guaranteed to be unique and are safe to be used for joins
+                data_mod = pd.concat(
+                    dfs,
+                    join="outer",
+                    axis=axis,
+                    sort=False,
                 )
-                for m, a in self.mod.items()
-            ]
-            data_mod = pd.concat(
-                dfs,
-                join="outer",
-                axis=axis,
-                sort=False,
-            )
 
-        # pd.concat wrecks the ordering when doing an outer join with a MultiIndex and different data frame shapes
-        if axis == 1:
-            newidx = (
-                reduce(lambda x, y: x.union(y, sort=False), (df.index for df in dfs))
-                .to_frame()
-                .reset_index(level=1, drop=True)
-            )
-            globalidx = data_global.index.get_level_values(0)
-            mask = globalidx.isin(newidx.iloc[:, 0])
-            if len(mask) > 0:
-                negativemask = ~newidx.index.get_level_values(0).isin(globalidx)
-                newidx = pd.MultiIndex.from_frame(
-                    pd.concat(
-                        [newidx.loc[globalidx[mask], :], newidx.iloc[negativemask, :]], axis=0
+                data_common = pd.concat(
+                    [
```

Comment:
also duplicated, can be moved outside the main `if` branch together with `columns_common =...`